markdown-mode.el

```diff
@@ -2599,30 +2633,60 @@ analysis."
         (set-match-data (list beg (point)))
         t))))
 
-(defun markdown-match-generic-metadata (regexp last)
+(defun markdown-get-match-boundaries (start-header end-header last &optional pos)
+  (save-excursion
+    (goto-char (if pos pos (point-min)))
+    (cl-loop
+     with cur-result = nil
+     while (and (< (point) last)
+                (re-search-forward (or start-header "\\`") last t)
+                (progn
+                  (setq cur-result (match-data))
+                  (re-search-forward
+                   (or end-header "\n\n\\|\n\\'\\|\\'") nil t)))
```

Comment:
Following expressions are loop invariants. They can be moved outside the loop.
-  `(or start-header "\\`")`
- `(or end-header "\n\n\\|\n\\'\\|\\'")`